modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItSecondaryIndexMultiRangeScanTest.java

```diff
@@ -0,0 +1,362 @@
+                .returns(5, "5", 25)
+                .check();
+    }
+
+    @ParameterizedTest(name = TEST_DISPLAY_NAME)
+    @MethodSource("testParameters")
+    public void testNulls(boolean useDynamicParameters, String direction) {
+        assertCondition(direction, useDynamicParameters, "c1 IS NULL AND c2 <= %s", "1")
+                .returns(null, "1", 0)
+                .check();
+
+        assertCondition(direction, useDynamicParameters, "(c1 IS NULL OR c1 = %s) AND c2 = %s AND c3 in (%s, %s)",
+                3, "1", 0, 3)
+                .returns(null, "1", 0)
+                .returns(3, "1", 3)
+                .check();
+
+        assertCondition(direction, useDynamicParameters, "(c1 IS NULL OR c1 < %s) AND c2 = %s AND c3 in (%s, %s)",
+                3, "1", 0, 1)
+                .returns(null, "1", 0)
+                .returns(1, "1", 1)
+                .check();
+
+        assertCondition(direction, useDynamicParameters, "(c1 IS NULL OR c1 > %s) AND c2 = %s AND c3 in (%s, %s)",
+                3, "5", 0, 25)
+                .returns(null, "5", 0)
+                .returns(5, "5", 25)
+                .check();
+
+        assertCondition(direction, useDynamicParameters, "c1 IS NOT NULL AND c2 IS NULL")
+                .returns(0, null, 0)
+                .check();
+
+        assertCondition(direction, useDynamicParameters, "c1 IN(NULL, %s) AND c2 = %s", 1, "1")
+                .returns(1, "1", 1)
+                .check();
+    }
+
+    /** Tests not supported range index scan conditions. */
+    @ParameterizedTest(name = "dynamicParams={0}")
+    @ValueSource(booleans = {true, false})
+    public void testNot(boolean useDynamicParameters) {
+        assertQuery(useDynamicParameters, "SELECT * FROM test_asc WHERE c1 <> %s AND c3 = %s", 1, 6)
+                .matches(QueryChecker.containsTableScan("PUBLIC", "TEST_ASC")) // Can't use index scan.
+                .returns(2, "3", 6)
+                .returns(3, "2", 6)
+                .check();
+
+        assertQuery(useDynamicParameters, "SELECT * FROM test_asc WHERE c1 NOT IN (%s, %s, %s) AND c2 NOT IN (%s, %s, %s)",
+                1, 2, 5, 1, 2, 5)
+                .matches(QueryChecker.containsTableScan("PUBLIC", "TEST_ASC")) // Can't use index scan.
+                .returns(3, "3", 9)
+                .returns(3, "4", 12)
+                .returns(4, "3", 12)
+                .returns(4, "4", 16)
+                .check();
+    }
+
+    /** Test correct index ordering without additional sorting. */
+    @ParameterizedTest(name = "dynamicParams={0}")
+    @ValueSource(booleans = {true, false})
+    public void testOrdering(boolean useDynamicParameters) {
+        assertQuery(useDynamicParameters, "SELECT * FROM test_asc WHERE c1 IN (%s, %s) AND c2 IN (%s, %s) ORDER BY c1, c2, c3",
+                3, 2, "3", "2")
+                .matches(CoreMatchers.not(QueryChecker.containsSubPlan("Sort"))) // Don't require additional sorting.
```

Comment:
These tests reply on index usage, why not to use USE_INDEX hint directly?

Reply:
I meant `FORCE_INDEX`.

Reply:
Done